### PR TITLE
add session_factory_from_settings and includeme

### PR DESCRIPTION
**Pull request by silenius**
Add session_factory_from_settings and includeme

### CODE REVIEW

File: pyramid_nacl_session/utils.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+            if k.startswith(prefix):
+                option_name = k[len(prefix):]
+
+                if option_name in bool_options:
+                    v = asbool(v)
+                elif option_name == 'serializer':
+                    if v in serializers:
+                        v = serializers[v]()
+                elif option_name == 'secret':
+                    v = unhexlify(v)
```
Comment (mmerickel) [inline pyramid_nacl_session/utils.py]: I sort of think the secret should just be treated as text. At least I haven't seen precedent in other libraries for requiring it to be a hexlify'd string.

Comment (silenius) [reply pyramid_nacl_session/utils.py]: secret must be a 32-byte string

File: .travis.yml [hunk 2]
```diff
@@ -4,16 +4,16 @@ sudo: false
...
-        - python: 3.3
-          env: TOXENV=py33
         - python: 3.4
           env: TOXENV=py34
         - python: 3.5
           env: TOXENV=py35
+        - python: 3.6
+          env: TOXENV=py36
+        - python: 3.7
+          env: TOXENV=py37
```
Comment (stevepiercy) [inline .travis.yml:14]: Please add Python 3.8.

File: docs/usage.rst [hunk 3]
```diff
@@ -1,34 +1,38 @@
...
-Generating a Random Secret
---------------------------
+Installation
+------------
 
-First, you need to arrange to have a 32-byte random secret available:
+Install using setuptools, e.g. (within a virtualenv)::
 
-.. code-block:: python
+  $ pip install pyramid_nacl_session
```
Comment (stevepiercy) [inline docs/usage.rst]: I think that installation of this package is seldom done by `pip install pyramid_nacl_session`, but usually by requiring it in `setup.py` or other methods. I think that we would better serve users by being vague about how to install this package ("Install pyramid_nacl_session via your preferred method.") or remove this section altogether.

However if you wish to preserve this section, then it should be:

```
Install using pip.

.. code-block:: bash

    pip install pyramid_nacl_session
```

File: docs/usage.rst [hunk 4]
```diff
@@ -1,34 +1,38 @@
...
+``config.include`` mechanism to include it into your Pyramid project's 
+configuration. In your Pyramid project's ``__init__.py``:
 
-Configure a Session Factory
----------------------------
+.. code-block:: python
 
-Use the generated secret to configure a session factory:
+   config = Configurator(.....)
+   config.include('pyramid_nacl_session')
```
Comment (stevepiercy) [inline docs/usage.rst]: Five dots is not syntactically correct and looks weird. An actual example would be fine, and better to use a context manager with `with`.

```
    with Configurator(settings=settings) as config:
        config.include('pyramid_nacl_session')
        # ... more config.includes
        config.scan()
```

File: docs/usage.rst [hunk 5]
```diff
@@ -1,34 +1,38 @@
...
 
-.. code-block:: python
+Alternately, instead of using the Configurator's ``include`` method, you can
+activate :mod:`pyramid_nacl_session` by changing your application's ``.ini`` 
+file, use the following line:
```
Comment (stevepiercy) [inline docs/usage.rst]:
```suggestion
file, using the following line:
```

File: docs/usage.rst [hunk 6]
```diff
@@ -1,34 +1,38 @@
...
-       config.set_session_factory(factory)
+   pyramid.includes = pyramid_nacl_session
 
-Sharing the Secret Across Instances
------------------------------------
+Usage
+-----
 
+Generate a secret key
+`````````````````````
```
Comment (stevepiercy) [inline docs/usage.rst]: Please use heading underlines consistent with the levels in our [Style Guide](https://docs.pylonsproject.org/projects/docs-style-guide/#headings) and adopted from [Python](https://devguide.python.org/documenting/#sections). In this case, `^`.

File: docs/usage.rst [hunk 7]
```diff
@@ -42,21 +46,57 @@ Cut-and-paste into your config file:
...
 .. code-block:: ini
 
-   yourapp.session_secret =
+   session.secret =
       840aaafdc36f067fbad9baf006efc0f672b86ab0dcb6a3e43ecc1f9d760915e5
 
-Parse the secret from the configuration, de-hexlify it, and use it to
-construct and register a session factory:
+Session management
+``````````````````
```
Comment (stevepiercy) [inline docs/usage.rst]: Use `^` underline.

File: docs/usage.rst [hunk 8]
```diff
@@ -42,21 +46,57 @@ Cut-and-paste into your config file:
...
+If you have included pyramid_nacl_session in your Pyramid project's configuration
```
Comment (stevepiercy) [inline docs/usage.rst]:
```suggestion
If you have included :mod:`pyramid_nacl_session` in your Pyramid project's configuration
```

File: docs/usage.rst [hunk 9]
```diff
@@ -1,34 +1,38 @@
...
 
-.. code-block:: python
+  $ pip install pyramid_nacl_session
 
-   from pyramid_nacl_session import generate_secret
+Setup
+-----
 
-   SECRET = generate_secret(as_hex=False)
+Once :mod:`pyramid_nacl_session` is installed, you typically use the
```
Comment (stevepiercy) [inline docs/usage.rst:7]: When making changes to the docs, [please use one sentence per line](https://docs.pylonsproject.org/projects/docs-style-guide/#use-of-whitespace).

File: docs/usage.rst [hunk 10]
```diff
@@ -42,21 +46,57 @@ Cut-and-paste into your config file:
...
+``.ini`` file.  For example:
 
-.. code-block:: python
+.. code-block:: ini
 
-   import binascii
-   from pyramid.session import BaseCookieSessionFactory
-   from pyramid_nacl_session import EncryptedCookieSessionFactory
+   [app:myapp]
+   .. other settings ..
```
Comment (stevepiercy) [inline docs/usage.rst]:
```suggestion
   # other settings
```

File: docs/usage.rst [hunk 11]
```diff
@@ -42,21 +46,57 @@ Cut-and-paste into your config file:
...
+   def app(global_config, **settings):
+       """ This function returns a WSGI application.
+       
+       It is usually called by the PasteDeploy framework during 
+       ``paster serve``.
+       """
+       session_factory = session_factory_from_settings(settings)
+       config = Configurator(root_factory=get_root, settings=settings)
+       config.set_session_factory(session_factory)
+       # ... other configuration stuff...
```
Comment (stevepiercy) [inline docs/usage.rst]: Please see above example suggestion that uses a context manager and adapt accordingly.

File: pyramid_nacl_session/utils.py [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+from pyramid.session import PickleSerializer
+
+from pyramid.settings import asbool
+
+from .session import EncryptedCookieSessionFactory
+
+
+def session_factory_from_settings(settings):
+    """ Return a Pyramid session factory using PyNaCl session settings
+    supplied from a Paste configuration file"""
```
Comment (stevepiercy) [inline pyramid_nacl_session/utils.py]: Nitpick: I prefer multi-line docstrings to have `"""` on their own separate lines. Nonetheless, docstrings must follow English punctuation rules.

```
    """
    Return a Pyramid session factory using PyNaCl session settings
    supplied from a Paste configuration file.
    """
```

File: pyramid_nacl_session/utils.py [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+    prefixes = ('session.', 'pynacl.session.')
+    options = {}
+    bool_options = {'secure', 'httponly', 'set_on_exception'}
+    serializers = {
+        'json': JSONSerializer,
+        'pickle': PickleSerializer
+    }
+
+    # Pull out any config args meant for PyNaCl session. if there are any
```
Comment (stevepiercy) [inline pyramid_nacl_session/utils.py]:
```suggestion
    # Pull out any config args meant for PyNaCl session, if there are any.
```

File: tox.ini [hunk 14]
```diff
@@ -1,18 +1,18 @@
 [tox]
 envlist =
 #   py26,py27,py33,py34,py35,pypy,pypy3,
-    py26,py27,py33,py34,py35,pypy
+    py27,py34,py35,py36,py37,pypy
```
Comment (stevepiercy) [inline tox.ini]: Add py38 here, too.

File: tox.ini [hunk 15]
```diff
@@ -1,18 +1,18 @@
...
 # Most of these are defaults but if you specify any you can't fall back
 # to defaults for others.
 basepython =
-    py26: python2.6
     py27: python2.7
-    py33: python3.3
     py34: python3.4
     py35: python3.5
+    py36: python3.6
+    py37: python3.7
```
Comment (stevepiercy) [inline tox.ini:13]: ...and here.

**Review by stevepiercy — CHANGES_REQUESTED** [review]

File: docs/usage.rst [hunk 16]
```diff
@@ -50,53 +40,39 @@ Cut-and-paste into your config file:
...
-       # ... other configuration stuff...
-       return config.make_wsgi_app()
+      It is usually called by the PasteDeploy framework during 
+      ``paster serve``.
+      """
+      session_factory = session_factory_from_settings(settings)
+      with Configurator(settings=settings) as config:
+         config.set_session_factory(session_factory)
+         # other configuration stuff
+         return config.make_wsgi_app()
```
Comment (stevepiercy) [inline docs/usage.rst]: All the above lines 65-78 are not correctly indented. Please use 4 spaces for reStructuredText and Python code samples.

Also please check other indentation earlier in this file to ensure it is 4 spaces.

Note, PyCharm is notorious for inheriting the first indentation as its default for the document which will cause Python code samples to be incorrectly spaced. Keeping all indents to 4 spaces avoids that issue. Other editors might exhibit the same behavior.

File: pyramid_nacl_session/session.py [hunk 17]
```diff
@@ -1,22 +1,26 @@
...
+        max_age=None,
+        path='/',
+        domain=None,
+        secure=False,
+        httponly=False,
+        timeout=1200,
+        reissue_time=0,
+        set_on_exception=True,
+        serializer=None,
 ):
```
Comment (stevepiercy) [inline pyramid_nacl_session/session.py]: Side note, if you use [Black](https://github.com/psf/black) for code formatting, I would not be opposed. We use it in Pyramid now.

File: tox.ini [hunk 18]
```diff
@@ -1,7 +1,7 @@
 [tox]
 envlist =
 #   py26,py27,py33,py34,py35,pypy,pypy3,
-    py27,py34,py35,py36,py37,pypy
+    py27,py34,py35,py36,py37,py38,pypy
```
Comment (stevepiercy) [inline tox.ini]: I missed a couple things on the first pass.

[Python 3.4 is now EOLed](https://www.python.org/dev/peps/pep-0429/#release-schedule). Suggest its removal as well, both here and below.

Also the commented line 3 should be dropped.

File: .travis.yml [hunk 19]
```diff
@@ -14,6 +14,8 @@ matrix:
           env: TOXENV=py36
```
Comment (stevepiercy) [inline .travis.yml:12]: I missed this on my first pass. [Python 3.4 is now EOLed](https://www.python.org/dev/peps/pep-0429/#release-schedule). Suggest its removal as well, both here and below.

File: setup.py [hunk 20]
```diff
@@ -11,7 +15,7 @@
 tests_require = ['WebTest']
 
 setup(name='pyramid_nacl_session',
-      version='0.4.dev0',
+      version=__version__,
```
Comment (stevepiercy) [inline setup.py]: Please update the trove classifiers below to support Python 3.5, 3.6, 3.7, 3.8.

**Review by stevepiercy — CHANGES_REQUESTED** [review]
All your changes are good. I missed a couple of things on my first pass. One more round, and then let's get a maintainer to review. Thank you!

File: docs/usage.rst (hunk 16 above)
Comment (silenius) [reply docs/usage.rst]: > Other editors might exhibit the same behavior.

indeed, same in Neovim :smiling_imp:

File: pyramid_nacl_session/session.py (hunk 17 above)
Comment (silenius) [reply pyramid_nacl_session/session.py]: :heavy_check_mark:

**Review by stevepiercy — APPROVED** [review]
Thank you for your changes!

Next we need to figure out why Travis fails. I have no clue how to fix it.

Alternatively we could switch to GitHub Actions.

**Comment by stevepiercy** [comment]
Er, I meant, I have no idea why the pypy build fails.

We need more test coverage for the changes, and that should take care of that one failure.

**Comment by silenius** [comment]
I'll add some tests

**Comment by silenius** [comment]
I've added some tests, (`utils.py` has now a coverage of 100%), but it looks like there is an issue with the PyPy instance, and some coverage issue for `__init__.py` ..

**Comment by stevepiercy** [comment]
I have no idea how to resolve those remaining issues. Maybe a maintainer has an idea?

**Comment by silenius** [comment]
@mmerickel : I've readded the `unhexlify` in 44c2a6c otherwise we cannot use the string returned by `bin/print_secret` (because of https://github.com/Pylons/pyramid_nacl_session/blob/master/pyramid_nacl_session/serializer.py#L25-L28) ..

**Comment by silenius** [comment]
@tseaver can you have a look once you have a little time? :relaxed:

File: CHANGES.rst [hunk 21]
```diff
@@ -4,7 +4,9 @@ Changelog
 0.4 (unreleased)
 ----------------
 
-- TBD
+- Add session_factory_from_settings
+
+- Add includeme
```
Comment (tseaver) [inline CHANGES.rst]: Please note the removed / added support for Python version here as well.  I'd actually prefer to review those changes in a separate PR, but I understand how you go here.

File: docs/index.rst [hunk 22]
```diff
@@ -1,20 +1,18 @@
 :mod:`pyramid_nacl_session`
 ===========================
 
-This Pyramid add-on defines an encrypting, pickle-based cookie serializer,
-using ``PyNaCl`` [1] to generate the symmetric encryption for the cookie state.
-
+This Pyramid add-on defines an encrypting, pickle-based cookie serializer, using ``PyNaCl`` [1] to generate the symmetric encryption for the cookie state.
```
Comment (tseaver) [inline docs/index.rst]: Please wrap at 80 characters.

File: docs/usage.rst [hunk 23]
```diff
@@ -1,62 +1,78 @@
 Using ``pyramid_nacl_session``
 ==============================
 
-Generating a Random Secret
---------------------------
+Setup
+-----
 
-First, you need to arrange to have a 32-byte random secret available:
+Once :mod:`pyramid_nacl_session` is installed, you typically use the ``config.include`` mechanism to include it into your Pyramid project's configuration.
```
Comment (tseaver) [inline docs/usage.rst]: Here and below, please wrap at 80 characters.

File: pyramid_nacl_session/__init__.py [hunk 24]
```diff
@@ -1,12 +1,12 @@
...
-    """
+# -*- coding: utf-8 -*-
 
+# flake8: noqa
 from .serializer import EncryptedSerializer
-_export(EncryptedSerializer)
 from .session import EncryptedCookieSessionFactory
-_export(EncryptedCookieSessionFactory)
 from .scripts import generate_secret
-_export(generate_secret)
```
Comment (tseaver) [inline pyramid_nacl_session/__init__.py:12]: Rather than suppressing all `flake8` checks, please add the exported names to `__all__` to appease `flake8` (rather than the `_export` hack).

File: pyramid_nacl_session/scripts.py [hunk 25]
```diff
@@ -27,22 +27,23 @@ def print_secret():  # pragma: NO COVER
     The secret is a 32-byte random string, encoded using
     :func:`binascii.hexlify`.
     """
+
```
Comment (tseaver) [inline pyramid_nacl_session/scripts.py:30]: Janitorial changes (e.g., blackening) should normally be done in a separate PR:  they make reviewing substantive changes much more difficult.

File: pyramid_nacl_session/serializer.py [hunk 26]
```diff
@@ -21,11 +21,13 @@ class EncryptedSerializer(object):
...
     """
+
     def __init__(self, secret, serializer=None):
         if len(secret) != SecretBox.KEY_SIZE:
             raise ValueError(
-                "Secret should be a random bytes string of length %d" %
-                SecretBox.KEY_SIZE)
+                "Secret should be a random bytes string of length %d"
+                % SecretBox.KEY_SIZE
+            )
```
Comment (tseaver) [inline pyramid_nacl_session/serializer.py:31]: Rather than a gratuitous re-wrap, please use the more modern `.format()` syntax instead, e.g.:

```python
                "Secret should be a random bytes string of length {}".format(
                    SecretBox.KEY_SIZE
```

File: pyramid_nacl_session/tests/test_session.py [hunk 27]
```diff
@@ -1,34 +1,115 @@
-from pyramid import testing
+# -*- coding: utf-8 -*-
+
 import unittest
 import webtest
+from pyramid import testing
+
+
+class SomeClass:
```
Comment (tseaver) [inline pyramid_nacl_session/tests/test_session.py]: Given that we are still supporting Python 2.7, use `object` for clarity:

```suggestion
class SomeClass(object):
```

**Review by tseaver — CHANGES_REQUESTED** [review]
In addition to my inline comments:

- The PyPy test failure is due to PyPy 7.1.1 (the one which supplies the 2.7.13 API) shipping an older `cffi` library:

```bash
$ /opt/pypy-7.1.1-linux_x86_64-portable/bin/virtualenv-pypy /tmp/foobar
New pypy executable in /tmp/foobar/bin/pypy
Installing setuptools, pip, wheel...
done.
$ /tmp/foobar/bin/pip list
DEPRECATION: A future version of pip will drop support for Python 2.7. More details about Python 2 support in pip, can be found at https://pip.pypa.io/en/latest/development/release-process/#python-2-support
Package    Version
---------- -------
cffi       1.12.3 
greenlet   0.4.13 
pip        19.3.1 
readline   6.2.4.1
setuptools 42.0.2 
wheel      0.33.6 
```

`PyNacl` needs `cffi >= 1.4.1`, but [doesn't enforce that on PyPy](https://github.com/pyca/pynacl/blob/4ff1d3d52bc298108f56b533ba3675670457f2ff/setup.py#L45-L53).

- The coverage failure is for the newly-added `includeme` function.

File: docs/index.rst (hunk 22 above)
Comment (silenius) [reply docs/index.rst]: ah, I thought documentation should not adhere to PEP8 maximum line length (https://docs.pylonsproject.org/projects/docs-style-guide/#use-of-whitespace)

File: pyramid_nacl_session/scripts.py (hunk 25 above)
Comment (silenius) [reply pyramid_nacl_session/scripts.py:30]: Ok sorry, I'll think about it next time..!

**Comment by silenius** [comment]
> The coverage failure is for the newly-added `includeme` function.

I just added a test for the newly-added `includeme` function.

**Comment by stevepiercy** [comment]
@tseaver for the PyPy failure, what should we do? File an issue or PR against PyNacl? Merge this PR as is?

File: pyramid_nacl_session/session.py [hunk 28]
```diff
@@ -1,15 +1,19 @@
+# -*- coding: utf-8 -*-
```
Comment (digitalresistor) [inline pyramid_nacl_session/session.py]: Please do not add this to Python files in this repository. The default is UTF-8 on Python 3 anyway.

File: pyramid_nacl_session/session.py [hunk 29]
```diff
@@ -1,15 +1,19 @@
...
+
 from pyramid.session import (
     BaseCookieSessionFactory,
     PickleSerializer,
 )
 
 from .serializer import EncryptedSerializer
 
+
+# pylint: disable=too-many-arguments,invalid-name
```
Comment (digitalresistor) [inline pyramid_nacl_session/session.py]: I don't believe we currently use pylint on this repository. Please do not add extra linter information.

File: pyramid_nacl_session/session.py [hunk 30]
```diff
@@ -93,10 +97,7 @@ def EncryptedCookieSessionFactory(
     if serializer is None:
         serializer = PickleSerializer()
 
-    encrypted_serializer = EncryptedSerializer(
-        secret,
-        serializer=serializer,
-    )
+    encrypted_serializer = EncryptedSerializer(secret, serializer=serializer,)
```
Comment (digitalresistor) [inline pyramid_nacl_session/session.py]:
```suggestion
    encrypted_serializer = EncryptedSerializer(secret, serializer=serializer)
```

File: pyramid_nacl_session/tests/test_session.py [hunk 31]
```diff
@@ -1,34 +1,135 @@
-from pyramid import testing
+# -*- coding: utf-8 -*-
```
Comment (digitalresistor) [inline pyramid_nacl_session/tests/test_session.py]: Please do not add unnecessary changes to this PR.

File: setup.py [hunk 32]
```diff
@@ -1,3 +1,7 @@
+# -*- coding: utf-8 -*-
+
+__version__ = '0.4.dev2'
```
Comment (digitalresistor) [inline setup.py]: Please do not move the version information into `__version__`, this is unnecessary and not a convention we use for Pylons Projects.

File: pyramid_nacl_session/utils.py [hunk 33]
```diff
@@ -0,0 +1,42 @@
...
+from .session import EncryptedCookieSessionFactory
+
+
+def session_factory_from_settings(settings):
+    """
+    Return a Pyramid session factory using PyNaCl session settings
+    supplied from a Paste configuration file.
+    """
+
+    prefixes = ("session.", "pynacl.session.")
```
Comment (digitalresistor) [inline pyramid_nacl_session/utils.py:17]: I am not a huge fan of using a `session.` prefix without adding support for this to other session libraries or potentially to Pyramid directly. It may cause confusion because its not supported for anything else.

File: pyramid_nacl_session/__init__.py [hunk 34]
```diff
@@ -1,12 +1,15 @@
-def _export(thing):
-    """ No-op; document that ``thing`` is an API.
-
-    Also, shuts pyflakes up about unused import.
-    """
+# -*- coding: utf-8 -*-
 
+# flake8: noqa
```
Comment (digitalresistor) [inline pyramid_nacl_session/__init__.py]: Please remove this unnecessary line.

**Review by digitalresistor — CHANGES_REQUESTED** [review]

File: pyramid_nacl_session/utils.py (hunk 33 above)
Comment (silenius) [reply pyramid_nacl_session/utils.py:17]: `pyramid_beaker` supports `session.` too

File: docs/index.rst [hunk 35]
```diff
@@ -1,20 +1,19 @@
...
 using ``PyNaCl`` [1] to generate the symmetric encryption for the cookie state.
 
-
 [1] https://pynacl.readthedocs.io/en/latest/secret/
 
 Contents:
 
 .. toctree::
-   :maxdepth: 2
+    :maxdepth: 2
```
Comment (digitalresistor) [inline docs/index.rst:12]: I didn't notice this before, but the 3 spaces vs the 4 is intentional. In restructured text documents you match the `.. ` spacing not the usual 4 spaces in Python code.

File: docs/index.rst [hunk 36]
```diff
@@ -1,20 +1,19 @@
 :mod:`pyramid_nacl_session`
 ===========================
 
-This Pyramid add-on defines an encrypting, pickle-based cookie serializer,
+This Pyramid add-on defines an encrypting, pickle-based cookie serializer,
```
Comment (digitalresistor) [inline docs/index.rst]: Space at the end of a line is likely to get stripped by the next person to edit the file (I have it set up to do it automatically on files since in most cases it's a mistake), it is likely better to add this to the beginning of the next line.

File: pyramid_nacl_session/utils.py (hunk 33 above)
Comment (digitalresistor) [reply pyramid_nacl_session/utils.py:17]: I'm a -0 on it, but I'll allow it.

File: setup.py [hunk 37]
```diff
@@ -11,7 +11,7 @@
 tests_require = ['WebTest']
 
 setup(name='pyramid_nacl_session',
-      version='0.4.dev0',
+      version='0.4',
```
Comment (digitalresistor) [inline setup.py]: Don't drop the `.dev0`, the person who does the release will modify this value when the time comes.

**Comment by digitalresistor** [comment]
Thank you @silenius! Just a couple of other comments and I will happily sign off on it. We can fix the pypy failure in a separate PR :-).

File: docs/index.rst (hunk 35 above)
Comment (stevepiercy) [reply docs/index.rst:12]: We want 4 spaces for consistency. Editors will incorrectly indent Python code in `code-block` because they interpret the first indent as the file's default.

Comment (digitalresistor) [reply docs/index.rst:12]: Oh, okay.

**Comment by silenius** [comment]
> Thank you @silenius! Just a couple of other comments and I will happily sign off on it. We can fix the pypy failure in a separate PR :-).

Thank you for reviewing this :+1:

**Comment by silenius** [comment]
One last thing, should we document somewhere (`CHANGES.rst` ?) the namespace change from `yourapp.session_secret` to `session.` ?

File: docs/usage.rst (hunk 23 above)
Comment (stevepiercy) [reply docs/usage.rst]: This request for change is inconsistent with https://docs.pylonsproject.org/projects/docs-style-guide/#use-of-whitespace

If we cannot agree on a standard as an organization, then we should revise this section in the guide, giving a consistent message to contributors.

**Review by digitalresistor — APPROVED** [review]

**Comment by silenius** [comment]
ping ? :smiley:

**Comment by digitalresistor** [comment]
Thank you @silenius. Made an executive decision here.
